File: client/lib/features/favorites/favorites_store.dart
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:shared_preferences/shared_preferences.dart';

/// Persistent set of favorited listing ids, exposed as a reactive provider.
class FavoritesStore extends StateNotifier<Set<String>> {
  FavoritesStore() : super(const {}) {
    _hydrate();
  }
  static const _key = 'twende.favorites';

  Future<void> _hydrate() async {
    final prefs = await SharedPreferences.getInstance();
    state = prefs.getStringList(_key)?.toSet() ?? const {};
  }

  Future<void> toggle(String id) async {
    final next = {...state};
    if (next.contains(id)) {
      next.remove(id);
    } else {
      next.add(id);
    }
    state = next;
    final prefs = await SharedPreferences.getInstance();
    await prefs.setStringList(_key, next.toList());
  }

  bool contains(String id) => state.contains(id);
}

final favoritesProvider =
    StateNotifierProvider<FavoritesStore, Set<String>>((_) => FavoritesStore());

File: client/TwendeZanzabar/TwendeZanzabar/Features/FavoritesStore.swift
import Combine
import Foundation

/// Persistent set of favorited listing ids.
@MainActor
final class FavoritesStore: ObservableObject {
    static let shared = FavoritesStore()

    private let key = "com.twende.favorites"
    @Published private(set) var ids: Set<String>

    init() {
        if let saved = UserDefaults.standard.array(forKey: key) as? [String] {
            self.ids = Set(saved)
        } else {
            self.ids = []
        }
    }

    func contains(_ id: String) -> Bool { ids.contains(id) }

    func toggle(_ id: String) {
        if ids.contains(id) { ids.remove(id) } else { ids.insert(id) }
        UserDefaults.standard.set(Array(ids), forKey: key)
    }
}
